server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java

```diff
@@ -90,20 +91,33 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
                     throw new IllegalStateException("cannot add duplicate task: " + a);
                 }, IdentityHashMap::new));
             LinkedHashSet<BatchedTask> newTasks = new LinkedHashSet<>(tasks);
-            tasksPerBatchingKey.merge(firstTask.batchingKey, newTasks, (existingTasks, updatedTasks) -> {
-                for (BatchedTask existing : existingTasks) {
-                    // check that there won't be two tasks with the same identity for the same batching key
-                    BatchedTask duplicateTask = tasksIdentity.get(existing.getTask());
-                    if (duplicateTask != null) {
-                        throw new IllegalStateException(
-                            "task ["
-                                + duplicateTask.describeTasks(Collections.singletonList(existing))
-                                + "] with source ["
-                                + duplicateTask.source
-                                + "] is already queued"
-                        );
+            taskIdentityPerBatchingKey.merge(
```

Comment:
Yes @amkhar , there can be few race condition between these two maps. Please check [https://github.com/opensearch-project/OpenSearch/pull/7305#discussion_r1178938486]. I have added those in that thread.

By looking at change, I agree this will change the behavior in that race condition.

But on second thought I am thinking we should check dedup logic with the tasks which are execution as well and not only which are in queue. We can check feasibility of it. 

Thoughts on it? @shwetathareja / @amkhar

Reply:
> But on second thought I am thinking we should check dedup logic with the tasks which are execution as well and not only which are in queue

Yes, we're checking dedup on tasks which are in execution. Here, we're also checking dedup for tasks which are already timed out. We could miss an important task if it was timed out, submitted to the TaskBatcher and we reject it saying duplicate.

Reply:
> We could miss an important task if it was timed out, submitted to the TaskBatcher and we reject it saying duplicate.

@amkhar Most likely this would never happen as in it is matter of timing what if it was enqueued when it was just about to timeout, then also it would have been rejected. So, caller doesn't really have control, when it will be exactly timed out. If the caller is enqueuing duplicate task, then they should handle it properly. This logic guarantees duplicates tasks will never be enqueued.

Reply:
Okay, we can go ahead with this then.